metronome/checkpointing/models/src/io/iohk/metronome/checkpointing/models/MerkleTree.scala

```diff
@@ -58,9 +61,124 @@ object MerkleTree {
       // Position of the leaf in the lowest level.
       leafIndex: Int,
       // Number of leaves in the lowest level.
+      // TODO: remove? not used when verifying the proof
       leafCount: Int,
       // Hashes of the "other" side of the tree, level by level,
       // starting from the lowest up to the highest.
-      siblingPath: IndexedSeq[MerkleTree.Hash]
+      siblingPath: IndexedSeq[Hash]
   )
+
+  def build(
+      elems: Iterable[Hash],
+      hashFn: (Hash, Hash) => Hash = concatThenKec256
+  ): MerkleTree = {
+    @tailrec
+    def buildTree(nodes: Seq[MerkleTree]): Seq[MerkleTree] = {
+      if (nodes.size <= 1)
+        nodes
+      else {
+        val paired = nodes.grouped(2).toSeq.map {
+          case Seq(a, b) =>
+            MerkleTree(hashFn(a.hash, b.hash), Some(a), Some(b))
+          case Seq(a) =>
+            // if the element has no pair we hash it with itself
+            MerkleTree(hashFn(a.hash, a.hash), Some(a))
+        }
+        buildTree(paired)
+      }
+    }
+
+    if (elems.isEmpty)
+      MerkleTree(Hash.empty)
+    else
+      buildTree(elems.toSeq.map(MerkleTree(_))).head
+  }
+
+  val concatThenKec256: (Hash, Hash) => Hash =
+    (a, b) => Hash(Keccak256(a ++ b))
+
+  def verifyProof(
+      proof: Proof,
+      root: Hash,
+      leaf: Hash,
+      hashFn: (Hash, Hash) => Hash = concatThenKec256
+  ): Boolean = {
+    def verify(currentHash: Hash, height: Int, siblings: Seq[Hash]): Hash = {
+      if (siblings.isEmpty)
+        currentHash
+      else {
+        val dir = navigateTree(height, proof.leafIndex)
+        val nextHash =
+          if (dir == 0) hashFn(currentHash, siblings.head)
+          else hashFn(siblings.head, currentHash)
+
+        verify(nextHash, height + 1, siblings.tail)
+      }
+    }
+
+    verify(leaf, 1, proof.siblingPath) == root
+  }
+
+  def generateProof(root: MerkleTree, elem: Hash): Option[Proof] = {
+    val leaves    = findLeaves(root)
+    val leafIndex = leaves.indexOf(elem)
+    val height    = leafCountToHeight(leaves.size)
```

Comment:
I was thinking that the concatenation of all leaves could be avoided by doing two things:
1. Store the size of the underlying Merkle tree along with the hash
2. Returning the index of the hash in the tree by traversal until its found and figuring out the index based on the path taken (building an index the opposite way as `navigateTree`).

It depends on the tree size I suppose and how many proofs we need to generate. In our case we'd only have to generate the proof for certain blocks, so we wouldn't have to store the trees where the sizes would take up space, and they are small anyway. 

Removing the `leafCount` from the `Proof` would make it even clearer that there's no need to concatenate the whole tree, since we don't need the full size, so let's do it!

Another observation is that if I know the hash of the item then then I probably know its index in the original tree as well: it's the position of the transaction in the body, so the index could be an input to the function. It could be the input _instead_ of the hash, so we don't have to re-hash the item, since the hash is available in the tree.

Maybe both are valid cases, if retrieving the block body is more costly and we already have the transaction at hand, ready to be hashed.

Reply:
So instead of `findLeaves` I ended up with functions `findHeight` and `findSize` functions which only traverse a single branch, `findElem` which return a single Int. I wouldn't expect our trees to be large enough to require such optimisation, but it was a fun exercise :)

Reply:
Really cool :)